components/Markets/ConditionalMarketOrderBook.tsx

```diff
@@ -10,15 +15,166 @@ export function ConditionalMarketOrderBook({
+
+          const _aggreateAsks = new Map();
+          _asks.forEach((order: { price: number, size: number }) => {
+            if (_aggreateAsks.get(order.price) === undefined) {
+              _aggreateAsks.set(order.price, order.size);
+            } else {
+              _aggreateAsks.set(order.price, _aggreateAsks.get(order.price) + order.size);
+            }
+          });
+          let __asks: any[][];
+          if (_aggreateAsks) {
+            __asks = Array.from(_aggreateAsks.entries()).map((side) => [
+              (side[0].toFixed(4)),
+              side[1],
+            ]);
+          } else {
+            return [[69, 0]];
+          }
+          console.log(ctx.slot);
+          setAsks(__asks);
+        } catch (err) {
+          console.error(err);
+        }
+      },
+      'processed'
+    );
+    provider.connection.onAccountChange(
+      market.bids,
+      (updatedAccountInfo, ctx) => {
+        try {
+          const leafNodes = openBookProgram.coder.accounts.decode('bookSide', updatedAccountInfo.data);
+          const leafNodesData = leafNodes.nodes.nodes.filter(
+            (x: AnyNode) => x.tag === 2,
+          );
+          const _bids: {
+            price: number;
+            size: number;
+          }[] = leafNodesData
+            .map((x: any) => {
+              const leafNode: LeafNode = openBookProgram.coder.types.decode(
+                'LeafNode',
+                Buffer.from([0, ...x.data]),
+              );
+              // const owner = leafNode.owner.toString();
+              const size = leafNode.quantity.toNumber();
+              const price = leafNode.key.shrn(64).toNumber() / 10_000;
+              // console.log(`\x1b[31mAsk\x1b[0m on ${account.market} proposal ${account.proposalId} by ${owner} on slot ${ctx.slot} for ${size} @ $${price}`);
+              return {
+                price,
+                size,
+              };
+            })
+            .sort((
+              a: { price: number, size: number },
+              b: { price: number, size: number }) => b.price - a.price);
+
+          const _aggreateBids = new Map();
+          _bids.forEach((order: { price: number, size: number }) => {
+            if (_aggreateBids.get(order.price) === undefined) {
+              _aggreateBids.set(order.price, order.size);
+            } else {
+              _aggreateBids.set(order.price, _aggreateBids.get(order.price) + order.size);
+            }
+          });
+          let __bids: any[][];
+          if (_aggreateBids) {
+            __bids = Array.from(_aggreateBids.entries()).map((side) => [
+              (side[0].toFixed(4)),
+              side[1],
+            ]);
+          } else {
+            return [[69, 0]];
+          }
+          console.log(ctx.slot);
+          setBids(__bids);
+        } catch (err) {
+          console.error(err);
+        }
+      },
+      'processed'
+    );
+    }
+  };
+
+  useEffect(() => {
+    listenOrderBook();
+  });
```

Comment:
jw, if we think this is optimal. no deps mean this just runs once on load?

then does `onAccountChange` on the program connection above work to respond to updates to pass/fail market orders?

Reply:
> then does `onAccountChange` on the program connection above work to respond to updates to pass/fail market orders?

It's a websocket subscription. So when those accounts update then it triggers. I'm not sure I've done ws in client side next / react, and then on top of that with web3 ws... so this likely needs to have some handler to check for connection etc. but testing last night it seemed to keep the connection alive the whole time.